### PR TITLE
Use StreamD by default, expose StreamK as separate module

### PR DESCRIPTION
**Pull request by harendra-kumar**
Basically remove depdency on GHC rewrite rules. The reason for separating StreamD and StreamK instead of using rewrite rules:

* Separate types provide better reasoning for the programmer about performance. Each type has its own pros and cons and the programmer can choose the best one based on the use case.
* rewrite rules are fragile, led to broken performance in the past dues to change in GHC.
* Rewrite rules lead to optimization problems, blocking fusion in some cases, specifically when combining multiple operations e.g. (filter . drop).
* Rewrite rules lead to problems when calling a function recursively. For example, the StreamD version of foldBreak cannot be used recursively when wrapped in rewrite rules because each recursive call adds a roundtrip conversion from D to K and back to D. We can use the StreamK versions of these though because the rewrite rule gets eliminated in that case.
* If we have a unified module, we need two different versions of several operations e.g. appendK and appendD, both are useful in different cases.

### CODE REVIEW

File: core/src/Streamly/Data/Stream.hs [hunk 1]
```diff
@@ -69,6 +51,31 @@
 -- "Data.List" like functions and many more powerful combinators to perform
 -- common programming tasks.
 --
+-- == Performance Notes
+--
+-- The 'Stream' type represents a stream by composing state as data which
+-- enables stream fusion. Stream fusion generates a tight loop without any
```
Comment (adithyaov) [inline core/src/Streamly/Data/Stream.hs:57]: This statement is not sitting well with me. `represents a stream by`

File: core/src/Streamly/Data/Stream/StreamDK.hs [hunk 2]
```diff
@@ -0,0 +1,505 @@
+-- |
+-- Module      : Streamly.Data.Stream.StreamDK
```
Comment (adithyaov) [inline core/src/Streamly/Data/Stream/StreamDK.hs:2]: I'm not reviewing this module.

File: core/src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs [hunk 3]
```diff
@@ -308,6 +326,12 @@ parseBreak (PRD.Parser pstep initial extract) stream@(Stream step state) = do
             PR.Error err ->
                 return (Left (ParseError err), StreamD.nil)
 
+-- | Parse a stream using the supplied 'Parser'.
+--
+{-# INLINE parseBreak #-}
+parseBreak :: Monad m => PR.Parser a m b -> Stream m a -> m (Either ParseError b, Stream m a)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:332]: This is an internal module but the naming inconsistency will be changed later right?
`D` is default
`K` is something additional

Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:332]: parseBreakD is using ParserD, parseBreak uses Parser. Both use StreamD.

File: core/src/Streamly/Internal/Data/Stream/StreamD/Nesting.hs [hunk 4]
```diff
@@ -396,40 +403,6 @@ roundRobin (Stream step1 state1) (Stream step2 state2) =
             Skip s -> Skip (InterleaveFirstOnly s)
             Stop -> Stop
 
-------------------------------------------------------------------------------
--- Zipping
-------------------------------------------------------------------------------
-
-{-# INLINE_NORMAL zipWithM #-}
-zipWithM :: Monad m
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Stream/StreamD/Nesting.hs:404]: This went to the lower level StreamD/Type.hs module.

**Comment by harendra-kumar** [comment]
Benchmarks:

```
Data.Stream(cpuTime)
Benchmark                                                                default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------ -------------- --------------------------
All.Data.Stream/o-n-stack.iterated.dropOne (n/10 x 10)                           204.99                   +7427.18
All.Data.Stream/o-n-stack.iterated.dropWhileTrue (n/10 x 10)                     220.69                   +1120.75
All.Data.Stream/o-n-stack.iterated.takeAll (n/10 x 10)                          4255.34                     +38.56
All.Data.Stream/o-1-space.mixedX4.take-map                                      6301.19                     +33.27
All.Data.Stream/o-n-stack.iterated.filterEven (n/10 x 10)                       1503.39                     +31.91
All.Data.Stream/o-1-space.grouping.refoldIterateM                                367.99                     +15.35
All.Data.Stream/o-1-space.indexingx4.indexedR                                   4529.11                     +12.83

All.Data.Stream/o-1-space.Applicative.(<*>) (sqrt n x sqrt n)                     45.40                     -11.39
All.Data.Stream/o-1-space.concat.concatMapPure (sqrt n of sqrt n)               1150.99                     -12.62
All.Data.Stream/o-1-space.Applicative.(*>) (sqrt n x sqrt n)                      45.59                     -13.68
All.Data.Stream/o-1-space.elimination.uncons                                     895.20                     -14.57
All.Data.Stream/o-1-space.mixed.foldl-map                                      12663.30                     -18.25
All.Data.Stream/o-1-space.exceptions/serial.retryAll                           17945.60                     -20.28
All.Data.Stream/o-1-space.mixed.sum-product-scan                               34914.80                     -27.32
All.Data.Stream/o-1-space.Monad.(>>=) (sqrt n x sqrt n)                         6840.91                     -28.98
All.Data.Stream/o-1-space.Monad.(>>=) (sqrt n x sqrt n) (filterAllIn)           7225.98                     -29.03
All.Data.Stream/o-1-space.Monad.(>>=) (sqrt n x sqrt n) (breakAfterSome)        7566.95                     -29.88
All.Data.Stream/o-1-space.mixedX4.take-scan                                    98100.30                     -32.72
All.Data.Stream/o-1-space.mixedX4.drop-scan                                    99590.10                     -33.99
All.Data.Stream/o-1-space.mixedX4.filter-scan                                  99547.10                     -41.17
All.Data.Stream/o-1-space.mixedX4.scan-map                                    104136.00                     -48.90
All.Data.Stream/o-1-space.Monad.(>>) (sqrt n x sqrt n)                            68.18                     -51.14
All.Data.Stream/o-1-space.mixed.scanl-map                                      24212.30                     -52.43
All.Data.Stream/o-n-stack.iterated.scanl' (quadratic) (n/100 x 100)           151613.00                     -67.50
All.Data.Stream/o-n-space.Monad.(>>=) (sqrt n x sqrt n) (toListSome)            1507.30                     -68.37
All.Data.Stream/o-n-stack.iterated.mapM (n/10 x 10)                             2707.56                     -80.78
All.Data.Stream/o-1-space.Monad.(>>=) (sqrt n x sqrt n) (filterAllOut)          4957.42                     -84.12
All.Data.Stream/o-n-space.Monad.(>>=) (sqrt n x sqrt n) (toList)               27233.90                     -88.53
All.Data.Stream/o-1-space.Monad.(>>=) (sqrt n x sqrt n) (filterSome)            5202.44                     -90.56
All.Data.Stream/o-1-space.concat.concatMapId (n of 1) (fromFoldable)            4078.74                     -98.77
All.Data.Stream/o-1-space.filteringX4.drop-all                                   855.65                    -189.49
All.Data.Stream/o-1-space.filteringX4.dropWhile-true                             974.83                    -266.96
All.Data.Stream/o-1-space.filteringX4.dropWhileM-true                           1053.51                    -281.93
All.Data.Stream/o-1-space.mixed.sum-product-fold                               50483.10                    -364.95
All.Data.Stream/o-1-space.foldable.min (ord)                                    1067.78                    -613.99
All.Data.Stream/o-1-space.filteringX4.findIndices                               1020.59                   -1462.44
All.Data.Stream/o-1-space.filteringX4.elemIndices                               1055.56                   -1516.29
All.Data.Stream/o-1-space.filteringX4.drop-one                                  5380.36                   -2048.16
All.Data.Stream/o-1-space.filteringX4.uniq                                      7330.50                   -2083.31
All.Data.Stream/o-1-space.filteringX4.dropWhile-false                           5456.47                   -2328.81
All.Data.Stream/o-1-space.filteringX4.filter-all-out                             933.51                   -2391.63
All.Data.Stream/o-1-space.mixedX4.take-drop                                     9342.08                   -2402.29
All.Data.Stream/o-1-space.filteringX4.filterM-all-out                            944.71                   -2407.87
All.Data.Stream/o-1-space.pipesX4.tee                                          67332.60                   -2429.55
All.Data.Stream/o-1-space.insertingX4.insertBy                                  7760.32                   -2533.03
All.Data.Stream/o-1-space.filteringX4.take-all                                  5389.01                   -3507.83
All.Data.Stream/o-1-space.mixedX4.filter-drop                                  11259.40                   -4124.27
All.Data.Stream/o-1-space.joining.serial (2,x/2)                                2854.52                   -7520.53
All.Data.Stream/o-1-space.mixedX4.filter-take                                  11680.20                   -7723.10
All.Data.Stream/o-1-space.hoisting.generalizeInner                              3980.59                   -7886.10
All.Data.Stream/o-1-space.elimination.length . generalizeInner                  3965.41                   -8189.70
All.Data.Stream/o-1-space.filteringX4.filter-even                               3177.66                   -8360.07
All.Data.Stream/o-1-space.filteringX4.filterM-even                              3394.61                   -8979.53
All.Data.Stream/o-1-space.filteringX4.takeWhileM-true                           4909.49                  -10264.18
All.Data.Stream/o-1-space.filteringX4.takeWhile-true                            5128.10                  -10786.97
All.Data.Stream/o-1-space.joining.serial (2,2,x/4)                              4109.35                  -10873.07
All.Data.Stream/o-1-space.mappingX4.trace                                       4602.03                  -12211.81
All.Data.Stream/o-1-space.mappingX4.mapM                                        4955.25                  -13086.43
All.Data.Stream/o-1-space.pipesX4.mapM                                          5086.12                  -13525.41
All.Data.Stream/o-1-space.filteringX4.deleteBy                                  5114.29                  -13579.99
All.Data.Stream/o-1-space.filteringX4.filter-all-in                             5557.04                  -14778.65
All.Data.Stream/o-1-space.filteringX4.filterM-all-in                            5809.11                  -15454.21
All.Data.Stream/o-1-space.pipesX4.compose                                       7268.30                  -19243.14
```

```
Data.Stream.Concurrent(cpuTime)
Benchmark                                                                             default(0)(ms) default(1) - default(0)(%)
------------------------------------------------------------------------------------- -------------- --------------------------
All.Data.Stream.Concurrent/o-1-space.joining.concat async (2 of n/2)                           10.92                      +9.23
All.Data.Stream.Concurrent/o-1-space.concat.concat . fmap (n of 1)                             44.88                      +8.19
All.Data.Stream.Concurrent/o-1-space.concat-foldable.foldMapWith (<>) (Stream)                 26.95                      +6.13
All.Data.Stream.Concurrent/o-1-space.mapping.mapM                                              27.99                      +3.46
All.Data.Stream.Concurrent/o-1-space.concat.parConcatMap (n of 1)                              44.11                      -0.07
All.Data.Stream.Concurrent/o-1-space.concat.parConcatMap (sqrt x of sqrt x)                    11.20                      -0.56
All.Data.Stream.Concurrent/o-n-heap.buffered.mkAsync                                           11.29                      -3.32
All.Data.Stream.Concurrent/o-1-space.monad-outer-product.toNullAp                              57.11                     -11.99
All.Data.Stream.Concurrent/o-1-space.concat-foldable.S.concatFoldableWith (<>) (List)          40.56                     -19.25
All.Data.Stream.Concurrent/o-1-space.concat-foldable.foldMapWith (<>) (List)                   37.95                     -31.60
All.Data.Stream.Concurrent/o-1-space.concat.parConcatMap (1 of n)                              32.47                     -47.00
All.Data.Stream.Concurrent/o-1-space.joining.async (2 of n/2)                                  17.77                     -56.54
```

```
Data.Stream.ConcurrentEager(cpuTime)
Benchmark                                                                                  default(0)(ms) default(1) - default(0)(%)
------------------------------------------------------------------------------------------ -------------- --------------------------
All.Data.Stream.ConcurrentEager/o-1-space.joining.concat async (2 of n/2)                           11.96                      +0.95
All.Data.Stream.ConcurrentEager/o-n-heap.buffered.mkAsync                                           11.32                      -1.70
All.Data.Stream.ConcurrentEager/o-1-space.concat-foldable.foldMapWith (<>) (List)                  693.64                      -3.96
All.Data.Stream.ConcurrentEager/o-1-space.monad-outer-product.toNullAp                            1397.07                      -4.20
All.Data.Stream.ConcurrentEager/o-1-space.concat.parConcatMap (sqrt x of sqrt x)                    13.66                      -6.34
All.Data.Stream.ConcurrentEager/o-1-space.concat.concat . fmap (n of 1)                            651.78                     -11.92
All.Data.Stream.ConcurrentEager/o-1-space.concat.parConcatMap (n of 1)                             766.90                     -24.74
All.Data.Stream.ConcurrentEager/o-1-space.concat-foldable.S.concatFoldableWith (<>) (List)         794.05                     -26.94
All.Data.Stream.ConcurrentEager/o-1-space.concat-foldable.foldMapWith (<>) (Stream)                929.52                     -36.22
All.Data.Stream.ConcurrentEager/o-1-space.joining.async (2 of n/2)                                  17.62                     -56.09
All.Data.Stream.ConcurrentEager/o-1-space.concat.parConcatMap (1 of n)                              33.57                     -57.07
All.Data.Stream.ConcurrentEager/o-1-space.mapping.mapM                                            1263.03                     -88.08
```

```
Data.Stream.ConcurrentInterleaved(cpuTime)
Benchmark                                                                                        default(0)(ms) default(1) - default(0)(%)
------------------------------------------------------------------------------------------------ -------------- --------------------------
All.Data.Stream.ConcurrentInterleaved/o-1-space.mapping.mapM                                              26.33                     +74.15
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat-foldable.S.concatFoldableWith (<>) (List)          30.56                     +65.05
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat-foldable.foldMapWith (<>) (Stream)                 28.62                     +59.56
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat-foldable.foldMapWith (<>) (List)                   28.54                     +58.77
All.Data.Stream.ConcurrentInterleaved/o-1-space.monad-outer-product.toNullAp                             111.12                     +17.20
All.Data.Stream.ConcurrentInterleaved/o-1-space.joining.concat async (2 of n/2)                           28.55                      +8.54
All.Data.Stream.ConcurrentInterleaved/o-n-heap.buffered.mkAsync                                           26.37                      +3.30
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat.parConcatMap (1 of n)                              56.94                      +1.65
All.Data.Stream.ConcurrentInterleaved/o-1-space.joining.async (2 of n/2)                                  28.15                      -2.74
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat.concat . fmap (n of 1)                             84.90                      -4.55
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat.parConcatMap (n of 1)                              87.05                     -12.93
All.Data.Stream.ConcurrentInterleaved/o-1-space.concat.parConcatMap (sqrt x of sqrt x)                    68.90                    -140.43
```

```
Data.Fold.Window(cpuTime)
Benchmark                                      default(0)(μs) default(1) - default(0)(%)
---------------------------------------------- -------------- --------------------------
All.scan.maximum (window size 30)                     2256.59                     +71.69
All.scan.range (window size 30)                       2462.80                     +51.59
All.scan.minimum (window size 1000)                  44456.30                     +43.66
All.scan.range (window size 1000)                    44291.50                     +42.68
All.scan.maximum (window size 1000)                  48291.10                     +40.05
All.scan.minimum (window size 30)                     2589.80                     +30.64
All.scan.minimum (window size 10)                     1377.49                      +9.08
All.scan.maximum (window size 10)                     1636.70                      +8.95
All.scan.range (window size 10)                       1463.31                      +5.70

```

**Pull request merged**